src/commands/multi_block/monitor.rs

```diff
@@ -237,93 +421,145 @@ where
+                // Ensure it's recorded locally if somehow missed before
+                if !submitted_rounds.contains(round).await {
+                    submitted_rounds.insert(round, n_pages, score, None).await;
+                }
                 return Ok(());
             }
             // Revert previous submission and submit the new one.
-            bail(listen, &client, signer.clone()).await?;
+            log::warn!(target: LOG_TARGET, "Mined solution score {:?} is better than submitted score {:?} for round {}. Bailing previous submission.", submitted_score, score, round);
+            bail(Listen::Finalized, &client, signer.clone()).await?;
+            // Clear local tracking as we are bailing
+            submitted_rounds
+                .remove(round, Some(RoundUntrackingEvent::Bailed))
+                .await;
         }
         CurrentSubmission::Incomplete(s) => {
-            // Submit the missing pages.
-            if s.score() == paged_raw_solution.score {
+            // Submit the missing pages if score matches.
+            if s.score() == submitted_score {
+                log::info!(target: LOG_TARGET, "Found incomplete submission for round {} with matching score {:?}. Submitting missing pages.", round, submitted_score);
                 let mut missing_pages = Vec::new();
-
                 for page in s.get_missing_pages() {
-                    let solution = paged_raw_solution.solution_pages[page as usize].clone();
-                    missing_pages.push((page, solution));
+                    // Clone only needed pages
+                    if let Some(solution_page) =
+                        paged_raw_solution.solution_pages.get(page as usize)
+                    {
+                        missing_pages.push((page, solution_page.clone()));
+                    } else {
+                        log::error!(target: LOG_TARGET, "Error: Mined solution missing expected page {} for round {}", page, round);
+                        // Handle error appropriately, maybe bail or return error
+                        return Err(Error::Other("Mined solution missing expected page".into()));
+                    }
                 }
 
-                // Use the appropriate submission method based on chunk_size
-                if chunk_size == 0 {
+                let submission_result = if chunk_size == 0 {
                     dynamic::inner_submit_pages_concurrent::<T>(
                         &client,
                         &signer,
                         missing_pages,
-                        listen,
+                        Listen::Finalized,
                     )
-                    .await?;
+                    .await
                 } else {
                     dynamic::inner_submit_pages_chunked::<T>(
                         &client,
                         &signer,
                         missing_pages,
-                        listen,
+                        Listen::Finalized,
```

Comment:
remove `Listen` as well from `dynamic::inner_submit_pages_chunked` it's only used by experimental multi block stuff